king_phisher/server/database/models.py

```diff
@@ -397,6 +398,95 @@ class Message(Base):
+	question_type = sqlalchemy.Column(sqlalchemy.String, nullable=False)
+	hint = sqlalchemy.Column(sqlalchemy.String)
+	url_reference = sqlalchemy.Column(sqlalchemy.String)
+	possible_answer_group_id = sqlalchemy.Column(sqlalchemy.Integer, nullable=False)
+	question_answer_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('question_answers.id'))
+	# relationships
+	submission_answer_link = sqlalchemy.orm.relationship('SubmissionAnswerLink', backref='module_questions', cascade='all, delete-orphan')
+
+@register_table
+class Module(TagMixIn, Base):
+	__tablename__ = 'modules'
+	created = sqlalchemy.Column(sqlalchemy.DateTime, nullable=False)
+	# relationships
+	module_question = sqlalchemy.orm.relationship('ModuleQuestion', backref='modules', cascade='all, delete-orphan')
+	test_module = sqlalchemy.orm.relationship('TestModule', backref='modules', cascade='all, delete-orphan')
+
+@register_table
+class PossibleAnswerLink(Base):
+	__repr_attributes__ = ('group_id', 'answer_id')
+	__tablename__ = 'possible_answer_link'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	question_answer_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('question_answers.id'), nullable=False)
+	group_id = sqlalchemy.Column(sqlalchemy.Integer)
+
+@register_table
+class QuestionAnswer(Base):
+	__repr_attributes__ = ('answer',)
+	__tablename__ = 'question_answers'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	answer = sqlalchemy.Column(sqlalchemy.String, nullable=False, unique=True)
+	case_sensitive = sqlalchemy.Column(sqlalchemy.Boolean, nullable=False, default=False)
+	# relationships
+	module_question = sqlalchemy.orm.relationship('ModuleQuestion', backref='question_answers', cascade='all, delete-orphan')
+	submission_answers = sqlalchemy.orm.relationship('SubmissionAnswersLink', backref='question_answers', cascade='all, delete-orphan')
+	possible_answer_link = sqlalchemy.orm.relationship('PossibleAnswerLink', backref='QuestionAnswer', cascade='all, delete-orphan')
+
+@register_table
+class SubmissionAnswerLink(Base):
+	__repr_attributes__ = ('test_sub_id', 'module_question_id', 'answer_id')
+	__tablename__ = 'submission_answers_link'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	test_sub_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_submissions.id'), nullable=False)
+	module_question_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('module_questions.id'), nullable=False)
+	question_answer_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('question_answers.id'), nullable=False)
+
+@register_table
+class Test(TagMixIn, Base):
+	__repr_attributes__ = ('name', 'minimum_score')
+	__tablename__ = 'tests'
+	created = sqlalchemy.Column(sqlalchemy.DateTime, nullable=False)
+	minimum_score = sqlalchemy.Column(sqlalchemy.Integer, nullable=False)
+	retakes = sqlalchemy.Column(sqlalchemy.Integer, default=0, nullable=False)
```

Comment:
Thinking about this, to simplify the evaluation logic, can you please rename this to `max_attempts` and have it default to 1 (and drop the nullable since you'll have a default).